pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java

```diff
@@ -321,17 +326,296 @@ private static void setValuesToMap(String keyColumnName, String valueColumnName,
+            if (dotNotation) {
+              keys.add(fieldPath + "." + i);
+            } else {
+              keys.add(path);
+            }
+          }
+        }
+      }
+    }
+    return keys;
+  }
+
+  /**
+   * Extract all keys recursively from a JSON object up to maxDepth
+   */
+  private static List<String> extractAllKeysRecursively(Object jsonObj, int maxDepth) {
+    return extractAllKeysRecursively(jsonObj, maxDepth, false);
+  }
+
+  /**
+   * Extract all keys recursively from a JSON object up to maxDepth with output format option
+   */
+  private static List<String> extractAllKeysRecursively(Object jsonObj, int maxDepth, boolean dotNotation) {
+    List<String> allKeys = new java.util.ArrayList<>();
+    try {
+      JsonNode node;
+      if (jsonObj instanceof String) {
+        node = JsonUtils.stringToJsonNode((String) jsonObj);
+      } else {
+        node = JsonUtils.stringToJsonNode(JsonUtils.objectToString(jsonObj));
+      }
+
+      extractKeysFromNode(node, "$", allKeys, maxDepth, 1, dotNotation);
+    } catch (Exception e) {
+      // Return empty list on error
+    }
+    return allKeys;
+  }
+
+  /**
+   * Recursively extract keys from a JsonNode
+   */
+  private static void extractKeysFromNode(JsonNode node, String currentPath, List<String> keys,
+      int maxDepth, int currentDepth) {
+    extractKeysFromNode(node, currentPath, keys, maxDepth, currentDepth, false);
+  }
+
+  /**
+   * Recursively extract keys from a JsonNode with output format option
+   */
+  private static void extractKeysFromNode(JsonNode node, String currentPath, List<String> keys,
+      int maxDepth, int currentDepth, boolean dotNotation) {
+    if (currentDepth > maxDepth) {
+      return;
+    }
+
+    if (node.isObject()) {
+      node.fieldNames().forEachRemaining(fieldName -> {
+        String newPath = currentPath + "['" + fieldName + "']";
+        String keyToAdd = dotNotation ? convertToDotNotation(newPath) : newPath;
+        keys.add(keyToAdd);
+
+        JsonNode childNode = node.get(fieldName);
+        if (currentDepth < maxDepth && (childNode.isObject() || childNode.isArray())) {
+          extractKeysFromNode(childNode, newPath, keys, maxDepth, currentDepth + 1, dotNotation);
+        }
+      });
+    } else if (node.isArray()) {
+      for (int i = 0; i < node.size(); i++) {
+        String newPath = currentPath + "[" + i + "]";
+        String keyToAdd = dotNotation ? convertToDotNotation(newPath) : newPath;
+        keys.add(keyToAdd);
+
+        JsonNode childNode = node.get(i);
+        if (currentDepth < maxDepth && (childNode.isObject() || childNode.isArray())) {
+          extractKeysFromNode(childNode, newPath, keys, maxDepth, currentDepth + 1, dotNotation);
+        }
+      }
+    }
+  }
+
+  /**
+   * Convert JsonPath format to dot notation
+   * Example: $['a']['b']['c'] -> a.b.c
+   * $[0]['name'] -> 0.name
+   */
+  private static String convertToDotNotation(String jsonPath) {
```

Comment:
This is super expensive, especially if we do it on a per row basis. We should avoid regexp replacement